### ISSUE TITLE
Feature Request: transformative mappings in expressions

### ISSUE THREAD

**Issue by tbateson** [nicklauslittle/horseshoe-j#9]
Assuming that ` `` ` is the syntax for expressions,
data:
```json
{ "a": "1", "b": 3, "c": [], "d": {} }
```
map construction in expressions:
```
{{<d}}{{.}}{{/d}}{{<f}}``a: b, c: {{#d}}{{>d}}{{/d}}``{{/f}} -> produces a partial that constructs a map for the given context
``e = {{>f}}`` -> prints `{ "1": 3, []: {} }`
```
array/list construction in expressions:
```
``ls = [1,2,3,]`` -> array constant
``e = [{{#ls}}{{.}},{{/ls}}]`` -> creates a copy of ls
```
Closing thoughts:
- Allow partials and mustache tags to be invoked inside expressions
- Expand mustache tags within expressions first, to allow for building up of expressions
- Allow for trailing `,` in array declarations
- Don't use `{` and `}` for map declarations to avoid confusion with the default mustache tags `{{` and `}}`
- Perhaps use `#[ ]` for set declarations like clojure?

**Comment by tbateson** [comment]
Quite a few features have been added since I first made this request, so I'd like to post an update as most of the syntax above doesn't mesh with it.

I'm mostly OK with how arrays have been implemented, however I don't see the square brackets as necessary. Currently the comma operator is used to chain multiple expressions, something I think would be better off saved for the `;` symbol. That way you could write `{{#1, 2, 3,}}` (which right now scopes into the literal `3`) instead of having to write (`{{#[1, 2, 3,]}}`). I'm more in favor of using semicolons to do what the comma operator currently does (i.e. `{{#1; 2; 3}}` would scope into the literal `3`). This way the above `{{#1, 2, 3,}}` would iterate over the values in the array.

Additionally maps could function very similarly, `{{#"a": 1, "b": 2,}}` would scope into the map with the key-value pairs `"a", 1` and `"b": 2`.

**Comment by tbateson** [comment]
After some further consideration, I'm thinking the best way to do this is actually how groovy does their arrays/maps. `{{#[1, 2, 3,]}}` is an array literal and `{{#[1: 2, 3: 4,]}}` is a map literal. This way you can embed lists in maps and maps in lists.

**Comment by nicklauslittle** [comment]
This should be implemented in f5d9b386955c542290148c5e5e2d67434b171de8.

Horseshoe now uses `;` for separating statements and the `,` operator is strictly for arrays / maps or as a method parameter separator. A trailing `;` is optional.

Horseshoe now uses `[]` or `{}` for array and map literals. The `{}` is now used for creating iterable arrays and maps. If commas are used in a context where they would otherwise not be allowed (e.g. `{{4, 5}}`, the result is interpreted as if it were wrapped in `{}` (iterable). These are considered "auto-converted" maps and arrays. The `[]` should be used anywhere iteration is not desired.
